### PR TITLE
MM-22782: Write fuzz testing for websockets

### PR DESCRIPTION
**Pull request by agnivade**
#### Summary

We add a Fuzz function that does some high-level fuzzing of the web_hub. This is not meant to be fully exhaustive, but just as an addendum to the existing load tests.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-22782

### CODE REVIEW

**Comment by agnivade** [comment]
> Would all this work from outside the repository, just by importing needed packages (app, model, testlib, etc.)?

Unfortunately, I tried that. But there seems to be some dependency with the tests and the package paths.

It might be doable, if the code does not depend on the test helper at all and creates all the teams and users from scratch. That might be a lot of work.

**Comment by streamer45** [comment]
> Unfortunately, I tried that. But there seems to be some dependency with the tests and the package paths.
> 
> It might be doable, if the code does not depend on the test helper at all and creates all the teams and users from scratch. That might be a lot of work.

I see. Do you think it would be the same amount of work/complexity to have all the fuzz files in a separate directory (package) in the server repository?

**Comment by agnivade** [comment]
Let me check that.

**Comment by agnivade** [comment]
Looks like that still needs work :laughing: 

```
panic: runtime error: invalid memory address or nil pointer dereference
github.com/mattermost/mattermost-server/v5/testlib.(*MainHelper).GetStore
github.com/mattermost/mattermost-server/v5/app.Setup
github.com/mattermost/mattermost-server/v5/fuzz.Fuzz.func1
sync.(*Once).doSlow
sync.(*Once).Do
github.com/mattermost/mattermost-server/v5/fuzz.Fuzz
go-fuzz-dep.Main
main.main
```

The issue seems to be that global variable `mainHelper`. For the store setup to work, that variable needs to be set correctly and from the same package. :(

File: app/webhub_fuzz.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	//	Assign userID, channelID, teamID randomly from respective byte indices
+	input.createUserID = userIDs[int(data[0])%len(userIDs)]
+	input.selectChannelID = channelIDs[int(data[1])%len(channelIDs)]
+	input.selectTeamID = teamIDs[int(data[2])%len(teamIDs)]
+	input.invalidateConnUserID = userIDs[int(data[3])%len(userIDs)]
+	input.updateConnUserID = userIDs[int(data[4])%len(userIDs)]
+	input.event = events[int(data[5])%len(events)]
+	data = data[6:]
+	input.attachment = make(map[string]interface{})
+	for len(data) >= 4 { // 2 bytes key, 2 bytes value
```
Comment (isacikgoz) [inline app/webhub_fuzz.go:136]: Out of curiosity: Is this the right approach to discard last bytes if we have 3 bytes remaining? I mean can we guarantee that we never get inputs like this?

File: app/webhub_fuzz.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+//
+// Since this requires help of the testing library, we have to duplicate some code
+// over here because go-fuzz cannot take code from _test.go files. It won't affect
+// the main build because it's behind a build tag.
+//
+// To run this:
+// 1. go get -u github.com/dvyukov/go-fuzz/go-fuzz github.com/dvyukov/go-fuzz/go-fuzz-build
+// 2. mv app/helper_test.go app/helper.go
+// (Also reduce the number of push notification workers to 1 to debug stack traces easily.)
+// 3. go-fuzz-build github.com/mattermost/mattermost-server/v5/app.
```
[the remark this answers is not in the document]
Comment (agnivade) [reply app/webhub_fuzz.go:44]: okay :)

File: app/webhub_fuzz.go (hunk 1 above)
Comment (agnivade) [reply app/webhub_fuzz.go:136]: We are just trying to craft an input for our program. No we can't guarantee, and we don't bother much. We will let go-fuzz figure out the next input to give based on our return codes and the coverage analysis that it does.

Comment (isacikgoz) [reply app/webhub_fuzz.go:136]: So basically we are treating `[10]byte{...}` and `[12]byte{...}` same.

Comment (agnivade) [reply app/webhub_fuzz.go:136]: Correct.

Comment (isacikgoz) [reply app/webhub_fuzz.go:136]: I see, thanks!

**Comment by agnivade** [comment]
@lindy65 - This just adds a fuzz test. There is no code added. Unless you want to test the test itself, there is nothing to be tested here :). Let me know.

Thanks.

**Pull request merged**